microsoft/PowerToys
# Functionality to detect Win32 apps which are installed, deleted or renamed while PowerToys is running

## Summary of the Pull Request
When a Win32 app is installed, deleted or renamed, PowerToys Run updates itself accordingly.


## PR Checklist
* [x] Applies to #3553


## Detailed Description of the Pull Request / Additional comments
The changes made in this PR are as follows -
* The `win32Storage` list has been modified to use a class, called `Win32ProgramRepository` which implements `ListRespository` and `IProgramRepository` just like in packaged applications. With this change all the items are stored in a concurrent dictionary and adding, deleting, indexing all programs has become easier. As concurrent dictionary is thread safe, we do not need locks while setting/deleting items.
* When a Win32 application is added, there can be a chance of duplicates. To avoid this situation, the `GetHashCode()` function of Win32 has been overridden so that the custom Win32 GetHashCode() function is used. This ensures that no duplicates are present in the list even when we actively add or delete apps.
* The `Win32ProgramRepository` and `Win32ProgramRepositoryHelper` classes have been added to watch certain locations and trigger `Created`, `Deleted` or `Renamed` events whenever such an operation is performed on a file. To prevent these events from being triggered for non app type files, we add `Filters`. These filters raise events only for those particular type of applications. [Filters](https://docs.microsoft.com/en-us/dotnet/api/system.io.filesystemwatcher.filters?view=netcore-3.1)
* The `FileSystemWatchers` filter out exe, lnk, appred-ms and url type files and call the `GetAppFromPath(path)` function which has been added to the Win32 class. This function returns the corresponding Win32 item which can be added to the list. 
* Wrappers and interfaces have been added to `FileSystemWatcher`, `System.IO.File` and `FileVersionInfo` classes so that they can be mocked and tested.
* `ShellLinkHelper` is implementing the interface `IShellLinkHelper` so that the Win32 shortcut programs (lnk) can be tested.

* For shortcut apps, `GetAppFromPath` doesn't give us the expected FullPath for the deleted and renamed application and therefore it must be treated in a separate way because the FullPath is set using the ShellLinkHelper and if the file does not exist we cannot extract the **exe FullPath** from the lnkFile.
* For `Deleted` apps since we have no more information to go on, we iterate through all the programs and remove the one which has the same lnkResolved path.
* For `Renamed` apps, we have more information about the existing app, so we construct what could have been the old app so that we can remove it from the list. This is done so that we have an app which has the same name, exe name and full path as the app which was added so that the hashcodes are the same.

## Validation Steps Performed
* Manually validated it.
* Added tests -
1.  To ensure that the we will not be introducing duplicates.
2. For each event {Created, Renamed, Deleted} and app type {exe, lnk, appred-ms, url} a test has been added to validate that the FileSystemWatcher work as expected.
* All the other program plugin tests pass as well.
![image](https://user-images.githubusercontent.com/28739210/87326150-368e3400-c4e7-11ea-84d5-3f7f14ad7be2.png)
